### ISSUE TITLE
TensorFlowInferenceInterface: OutOfMemoryError on Android SDK 16

### ISSUE THREAD

**Issue by Johnson145** [tensorflow/tensorflow#17763]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Android SDK 16 (4.1.2)
- **TensorFlow installed from (source or binary)**: compile 'org.tensorflow:tensorflow-android:+
- **TensorFlow version (use command below)**: latest
- **Python version**: NA
- **Bazel version (if compiling from source)**: NA
- **GCC/Compiler version (if compiling from source)**: NA
- **CUDA/cuDNN version**: NA
- **GPU model and memory**: NA
- **Exact command to reproduce**: NA

### Describe the problem
I found a strange bug using the _TensorFlowInferenceInterface_ on Android SDK 16. When closing the interface once (using `inferenceInterface.close()`), the next start (using `new TensorFlowInferenceInterface()`) will cause an OutOfMemoryError, even though there is enough free RAM available. In contrast, there isn't any problem when keeping the interface alive to reuse it. The latter implies wasting resources though. The only way to prevent the OOM seems to be the manual kill of the surrounding process (`android.os.Process.killProcess(android.os.Process.myPid());`).

Just to be clear: there is really enough free RAM available. I'm not only referring to the device's RAM in general, but to the individual Java heap of the Android app. I even outsourced the TensorFlow code to a separate process to ensure that it gets its own heap.

I found this behavior using an old Samsung Galaxy S2. Couldn't reproduce it on newer Android versions. As SDK 16 is already quite old, I don't know whether anybody is willing to dig deeper into this bug. However, it costs me a lot of time to isolate the problem and I hope the workaround (using a separate process and killing it manually after closing the interface) may at least save somebody else some time.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @petewarden: It has been 16 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @petewarden: It has been 31 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @petewarden: It has been 46 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jdduke** [comment]
What is the size of your model?

Closing the `TensorFlowInferenceInterface` isn't guaranteed to immediately free all the used memory from the JVM. Moreover, while your device might have free memory available, the JVM heap size is typically much smaller than the device's memory limit.

Can you try adding `android:largeHeap="true"` to your manifest? In the meantime, we'll see if there's any way for `TensorFlowInferenceInterface.close()` to be more aggressive about releasing any held resources, though again it's still not guaranteed to solve the problem due to the unpredictability of garbage collection.

**Comment by Johnson145** [comment]
Thanks for your reply.

> What is the size of your model?

I tried e.g. the Inception v3 model, which is about 90MB.

> while your device might have free memory available, the JVM heap size is typically much smaller than the device's memory limit.

As already stated above: Just to be clear: there is really enough free RAM available. I'm not only referring to the device's RAM in general, but to the individual Java heap of the Android app. I even outsourced the TensorFlow code to a separate process to ensure that it gets its own heap.

> Can you try adding android:largeHeap="true" to your manifest?

I'm already using this option, but the error still occurs from time to time.

> In the meantime, we'll see if there's any way for TensorFlowInferenceInterface.close() to be more aggressive about releasing any held resources,

Cool, thanks!

> Closing the TensorFlowInferenceInterface isn't guaranteed to immediately free all the used memory from the JVM.  [..] though again it's still not guaranteed to solve the problem due to the unpredictability of garbage collection.

Garbage collection shouldn't be a problem here, should it? I guess it doesn't matter whether the garbage collection has actually freed the resources before I do anything, as long as they have been "marked to be removed the next time the garbage collection runs". Just before the OOM is raised, garbage collection will be triggered (automatically) at least once. Just to be sure, I even tried to trigger it manually. However, it fails to actually free the required resources. Maybe it's about segmentation / fragmentation of the free RAM? Doesn't occur on newer Android versions though.

**Comment by jdduke** [comment]
I haven't been able to repro locally, but I'd be curious if a heap dump might help isolate the problem or identify a leak.

If possible, you might try 1) Call `inferenceInterface.close()`, 2) Force a GC, 3) capture a heap dump. From there I'd be happy to look at the dump to see if anything stands out.

**Comment by jdduke** [comment]
All that said, have you looked into using TensorFlow Lite? It can substantially reduce the required memory footprint for certain models, and allows zero-copy loads of the input model as well as tensor inputs. There are pre-trained models for Inception v3 (as well as various other common models @ https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/lite/g3doc/models.md).  See https://www.tensorflow.org/mobile/tflite/ for more documentation.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jdduke: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
